db.go

```diff
@@ -240,6 +245,107 @@ func newDBMetrics(db *DB, r prometheus.Registerer) *dbMetrics {
 	return m
 }
 
+// DBView provides APIs for read only operations on a database.
+type DBView struct {
+	logger     log.Logger
+	dir        string
+	registerer prometheus.Registerer
+}
+
+// NewDBView returns a new DB in the given directory only for read operations.
+func NewDBView(dir string, l log.Logger, r prometheus.Registerer) (*DBView, error) {
+	if _, err := os.Stat(dir); err != nil {
+		return nil, err
+	}
+
+	if l == nil {
+		l = log.NewNopLogger()
+	}
+
+	db := &DBView{
+		logger:     l,
+		dir:        dir,
+		registerer: r,
+	}
+
+	return db, nil
+}
+
+// Querier loads the wal and returns a new querier over the data partition for the given time range.
+// A goroutine must not handle more than one open Querier.
+func (dbRead *DBView) Querier(mint, maxt int64) (Querier, error) {
+	head, err := NewHead(dbRead.registerer, dbRead.logger, nil, 1)
+	if err != nil {
+		return nil, err
+	}
+
+	blocks, err := dbRead.Blocks()
+	if err != nil {
+		return nil, err
+	}
+
+	db := &DB{
+		dir:    dbRead.dir,
+		logger: dbRead.logger,
+		head:   head,
+		blocks: blocks,
+	}
+
+	// Set the min valid time for the ingested wal samples
+	// to be no lower than the maxt of the last block.
+	minValidTime := int64(math.MinInt64)
+	if len(blocks) > 0 {
+		minValidTime = blocks[len(blocks)-1].Meta().MaxTime
+	}
+
+	if err := db.head.Init(minValidTime); err != nil {
+		return nil, errors.Wrap(err, "read WAL")
+	}
+
+	return db.Querier(mint, maxt)
+}
+
+// Blocks returns the databases persisted blocks.
+func (dbRead *DBView) Blocks() ([]*Block, error) {
+	db := &DB{
```

Comment:
I would say lets leave as is and will refactor if/when needed.

Reply:
I disagree, and won't approve this PR with this. This is a blocker IMO

Reply:
I just double checked and It feels that the refactoring that is needed to completely detach it from the db struct might allow more unexpected bugs than keeping it detached. For example the shared `chunkenc.Pool` , checking if a block is not already open etc.

I am thinking if I add a test that will ensure that running the `DBReadOnly` funcs don't modify any files on disk should be sufficient to catch write operations.

Reply:
I doubt that this much of refactoring is needed. What exactly is required? Even my proposal had low impact and I was able to not use `db`: https://github.com/prometheus/tsdb/pull/589

> I am thinking if I add a test that will ensure that running the DBReadOnly funcs don't modify any files on disk should be sufficient to catch write operations.

This is kind of wrong and overeningeering IMO. Why we are trying to spend time on not needed tests instead of doing this properly? In the very worst scenario just copied needed method like `getBlock` etc

Reply:
I see both methods as equally wrong and right and tests will be needed sooner or later so I will sit on it few days and will revisit.

Reply:
I revisited this and although I think both methods have dissadvantages, I have refactored it to your suggestions and will try to revisit again at some point.